### PR TITLE
feat(dataflow): Add iterator based data flow library

### PR DESCRIPTION
**Pull request by GordonSmith**
Signed-off-by: Gordon Smith <GordonJSmith@gmail.com>

## Checklist:
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit message includes a "fixes" reference if appropriate.
  - [x] The commit is signed.
- [x] The change has been fully tested:
  - [ ] I have viewed all related gallery items
  - [ ] I have viewed all related dermatology items
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised new issues to address them separately

### CODE REVIEW

File: packages/dataflow/README.md [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+    filter(row => row.index % 2 === 0),
+    sort((l, r) => l.value - r.value),
+    first(3)
+);
+
+const outDataA = [...process([0.1, 0.2, 0.3, 0.7, 0.8, 0.9])];  // => [{index: 0, value: 0.1},{index: 2,value: 0.3}]
+const outDataB = [...process([0.1, 0.3, 0.4, 0.6, 0.8, 0.9])];  // => [{index: 0, value: 0.1},{index: 2,value: 0.4}]
+```
+
+**Remember** think to think of the data as flowing through the activities.
```
Comment (jbrundage) [inline packages/dataflow/README.md:64]: Seems like a grammatical issue here

Comment (GordonSmith) [reply packages/dataflow/README.md]: @jbrundage updated

File: packages/dataflow/src/__tests__/readme.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+chain([0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
+    process,
+    max(row => row.value)
+);  // => 4
+
+const process_2 = chain(
+    process,
+    min(row => row.value)
+);
+console.log(process_2([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]));  // => 0
```
Comment (jbrundage) [inline packages/dataflow/src/__tests__/readme.ts:59]: There are no tests in this file?

Comment (GordonSmith) [reply packages/dataflow/src/__tests__/readme.ts:59]: Yea - its just ensuring the JS in the readme is working - not testing a library function as such.

File: packages/dataflow/src/chain.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            for (const activity of items) {
+                // @ts-ignore
+                tail = activity(tail);
+            }
+            return tail;
+        }
+    }
+}
+
+//  TODO:  Switch to TS Vardic Types in 4.0
```
Comment (jbrundage) [inline packages/dataflow/src/chain.ts:29]: *Variadic

Comment (GordonSmith) [reply packages/dataflow/src/chain.ts]: @jbrundage Fixed

**Pull request merged**